src/cryptonote_core/miner.cpp

```diff
@@ -410,5 +464,344 @@ namespace cryptonote
+    m_is_background_mining_enabled = false;
+
+    if(!get_system_times(prev_total_time, prev_idle_time))
+    {
+      LOG_ERROR(__func__ << " : get_system_times call failed, background mining will NOT work!");
+      return false;
+    }
+    
+    while(!m_stop)
+    {
+        
+      try
+      {
+        // Commenting out the below since we're going with privatizing the bg mining enabled
+        // function, but I'll leave the code/comments here for anyone that wants to modify the
+        // patch in the future
+        // -------------------------------------------------------------------------------------
+        // All of this might be overkill if we just enforced some simple requirements
+        // about changing this variable before/after the miner starts, but I envision
+        // in the future a checkbox that you can tick on/off for background mining after
+        // you've clicked "start mining". There's still an issue here where if background
+        // mining is disabled when start is called, this thread is never created, and so
+        // enabling after does nothing, something I have to fix in the future. However,
+        // this should take care of the case where mining is started with bg-enabled, 
+        // and then the user decides to un-check background mining, and just do
+        // regular full-speed mining. I might just be over-doing it and thinking up 
+        // non-existant use-cases, so if the concensus is to simplify, we can remove all this fluff.
+        /*
+        while( !m_is_background_mining_enabled )
+        {
+          MGINFO(__func__ << " : background mining is disabled, waiting until enabled!");
+          boost::unique_lock<boost::mutex> enabled_lock( m_is_background_mining_enabled_mutex );        
+          m_is_background_mining_enabled_cond.wait( enabled_lock );
+        } 
+        */       
+        
+        // If we're already mining, then sleep for the miner monitor interval.
+        // If we're NOT mining, then sleep for the idle monitor interval
+        boost::this_thread::sleep_for( 
+          m_is_background_mining_started ? 
+            boost::chrono::seconds( BACKGROUND_MINING_MINER_MONITOR_INVERVAL_IN_SECONDS ) :
+            boost::chrono::seconds( get_min_idle_seconds() ) );
+      }
+      catch(boost::thread_interrupted)
+      {
+        MGINFO(__func__ << " : background miner thread interrupted ");
+        continue; // if interrupted because stop called, loop should end ..
+      }
+
+      bool on_ac_power = ac_line_status();
 
+      if( m_is_background_mining_started )
+      {
+        // figure out if we need to stop, and monitor mining usage
+        
+        // If we get here, then previous values are initialized.
+        // Let's get some current data for comparison.
+
+        if(!get_system_times(current_total_time, current_idle_time))
+        {
+          LOG_ERROR(__func__ << " : get_system_times call failed");
```

Comment:
Adding __func__ everywhere seems unneeded. Logs already have file:line. If you want function too, you can set MONERO_LOG_FORMAT to have "%func" in addition to its default (see top of contrib/epee/src/mlog.cpp)